New rule: Device motion based changes to the content can also be created from the user interface (#1276)
act-rules/act-rules.github.io: pages/glossary/changes-in-content.md

---
title: Changes in content
key: changes-in-content
unambiguous: true
objective: true

> input_aspects:
>   - CSS styling
>   - DOM tree

Reviewer: ```suggestion
input_aspects:
  - Accessibility tree
  - Audio output
  - CSS styling
  - DOM tree
```

---

> Changes in the content of a [web page][] occur when at least one of the following occurs:

Reviewer: ```suggestion
A _change in the content_ of a [web page][] occurs when at least one of the following occurs:
```

> - **visible changes:** Visible changes in content occur when the rendered pixels change in any part of the document that is currently within the viewport or that can be brought into the viewport via scrolling; or
> - **accessibility tree changes:** Changes in the content of the accessibility tree occur when any state, property or event of a node representing an [accessible object][] of the tree changes, or when any node is inserted in, or removed from the tree; or
> - **audible changes:** Audible changes in content occur when the audio rendered by the [web page][] change.

Reviewer: ```suggestion
- **visible changes:** the rendered pixels change in any part of the document that is currently within the viewport or that can be brought into the viewport via scrolling; or
- **accessibility tree changes:** any state, property or event of a node representing an [accessible object][] of the accessibility tree changes, or any node is inserted in, or removed from the accessibility tree; or
- **audible changes:** the audio rendered by the [web page][] changes.
```

> - **audible changes:** Audible changes in content occur when the audio rendered by the [web page][] change.

Reviewer: "Viewport" and "accessibility tree" may need links to specs.

Author: We haven't been doing this in other rules, but why not :-)

> - **audible changes:** Audible changes in content occur when the audio rendered by the [web page][] change.

Reviewer: "The audio changes" => when my computer plays some music, the audio keeps changing (I'm not listening to _4'33''_ that often). I have a hunch that this is not the kind of changes you want to include.
OTOH, given that these changes in content are only considered as reaction to some action, this is probably OK (if an app requires me to move the phone to keep playing music, it definitely fails that rule…)

Author: You're right. The definition doesn't handle time-based media as it is. If the app allows me to move forward or backwards an audio track by moving the device then it should fail. But when the audio is playing the audio is always changing (unless it's playing 4'33'' :-) )
This also happens if it used to control a video. The rendered pixels are changing as part of the video playing (in addition to any audio the video has).

I'll try to rewrite, introducing a reference to compare the output against in order to identify changes.

---
title: Changes in content
key: changes-in-content
unambiguous: true
objective: true
input_aspects:
  - Accessibility tree
  - Audio output
  - CSS styling
  - DOM tree
---

> A _change in the content_ of a [web page][] caused by an event occurs when, by comparing the [web page][] before and after the event firing, at least one of the following occurs:

Reviewer: There is something in the causality that creates confusion. As in, there are two different concepts mixed here:
1. the page changes in one of the 3 ways indicated.
2. the page changes as a reaction to an event.

This does creates confusion in the assumptions. Notably:
* time-based media => it causes changes but not changes due to an event, hence the need to specify "if there's time based media, ignore it".
* changes only caused by the event => which is a bit weird as you pointed out, but becomes clearer and/or redundant if de-coupled from the even causing it. Currently, any "change in the content" has to be caused by an event (or at least perceived as such) which make the assumption that there are no changes not caused by event weird… because this use the other meaning of change…

Maybe it is worth splitting the definition in two ("changes" and "changes caused by an event") or at least to somehow streamline the difference between the two concepts.

Author: But this definition is about "change in the content of a web page caused by an event"

I don't think it helps to have a definition of just change (because change implies comparing something, which implies having to define what is going to be compared).

However, I do think it will help if we focus on dom events, given that this definition if being used in rules that deal with dom events. Hopefully, this will make it easier to address the issues.

- **visible changes:** the rendered pixels change in any part of the document that is currently within the [viewport][] or that can be brought into the [viewport][] via scrolling; or
- **accessibility tree changes:** any state, property or event of a node representing an [accessible object][] of the [accessibility tree][] changes, or any node is inserted in, or removed from the [accessibility tree][]; or

> - **audible changes:** the audio reproduced by the [web page][] changes.

Reviewer: I have completely missed that "render" was also used for audio in W3C lingo 😥 
I'm backing up on previous comment, "render" is the correct word to use here. Sorry about that 😓

Author: No worries!

**Assumptions:**

- If the [web page][] is rendering time-based media, rendered pixels and audio will be changing as part of the playback. The comparison in this instance should compare the pixels and audio that are rendered if the event is not fired, with the ones that are rendered if the event is fired.

> - For the purpose of this definition, it is assumed that the event causing the changes is the only origin of changes to the [web page][] and therefore the page does not change if there is no such event.

Reviewer: That assumption seems very strong in its current state.
If two different events create changes, the assumption means that they can't happen (almost) simultaneously.

If I come back to the scenario you describe in https://github.com/act-rules/act-rules.github.io/pull/1276#discussion_r440710851 then once the time based media start, anything else (such as tilting to go to next song) will fail this assumption (there is another origin of changes). This is more or less caught by she previous point (ignoring changes that would have occurred anyway).

We both seem uneasy with this assumption (and/or the combination with the time based media note above). So we probably need to do something about it… I am not sure that completely removing it is the way to go…

---
In this definition, "event" can be almost anything (not just a DOM event). So I agree that changes that happen will be caused by some event, possibly far away (e.g. playing a 10h video will cause changes for 10h).

We need a way to figure out that a given change is caused by a given event. This is currently the note about the time based media, but could apply to other changes (e.g. caused by an asynchronous request completing, this can potentially create change long after the event that caused it, unless we consider that the request completion is the "event" (and not sending the request)).

This does illustrate that there is some degree of subjectivity (hopefully not ambiguity) in the definition. "event" not being a DOM event creates place for interpretation, hence subjectivity. This is not a problem but the frontmatter should be updated accordingly.

---
I thing we're pretty solid on what a "change" is (i.e., objective, unambiguous). But not so on what an "event" is (if not DOM event) and consequently on the causality. Maybe trying to get a better definition of "event" will help us figure out problems with this assumption?

Author: I hope that you agree that fixing this to dom events solves the problem :-)

> For the purpose of this definition, it is assumed that the event causing the changes is the only origin of changes to the [web page][] and therefore the page does not change if there is no such event.

Reviewer: Wouldn't the second assumption de facto exclude time-based media? (thus making the first one useless)
"the page does not change if there is no such event" => no time based media is played.

Author: I don't think so. Take this scenario
1 - The user activates play and the web page starts playing audio in the background (there was a change in content but for the purpose of this rule the event that caused it did not originate from device motion)
2 - The user flicks the device to the right to move to the next track (there was another change in content - this time originated by a device motion event - even though the audio was already playing)

The first assumption kicks in in the second part of this scenario and it is what allows us to detect the change. 

To be honest, I'm still not totally convinced that we need the second one. I can't really come up with an example of something that will cause changes to the content without being event-based.

> - For the purpose of this definition, it is assumed that the changes happen instantly after the event firing and therefore the comparison between the page before and after the event firing can be made at any time after the firing of the event.

Reviewer: Also, maybe sort of assuming that the event itself is instantaneous 🤔 
But that might be part of what we call "event" here.
So, an asynchronous request could be two events (making the request, getting an answer), or 1 non-instantaneous (making he request), or maybe 4 (starting to send the request, finishing sending, starting to receive the answer, finishing it).

---
Let's say that I can (maybe by tilting the phone) send a request for a bunch of data. After some times, I start receiving that data. It takes some time to download everything because I'm on an old EDGE network 📶 (or just because there is a lot of data) Once download is done, the phone takes some time processing it (e.g., computing some aggregates). Once this processing is done, the page is updated instantaneously (or at least faster than I can notice).

How many events are there in this scenario? Which one is causing the change in the page? Does the rule apply (or rather requires an UI control)?

Author: This is, for me, the biggest issue with the current definition... even moving to dom events we can have a device motion event that triggers an asynchronous call that will resolve only after 1 minute (or whatever other duration).
The only other option I see, is using the "10 minute" workaround. What do you think @Jym77?

> For the purpose of this definition, it is assumed that the changes happen instantly after the event firing and therefore the comparison between the page before and after the event firing can be made at any time after the firing of the event.

Reviewer: I don't know whether this assumption will hold in practice (due to either micro-delays that can happen, or design choice to delay a bit the reaction).
But I guess it is OK given that this is a manual check and that I guess testers won't look at the result within the next micro-seconds…

Author: That was my reasoning. This is going to be a manual check, therefore we don't really need to concern oursevels with micro-delays.

> For the purpose of this definition, it is assumed that the changes happen instantly after the event firing and therefore the comparison between the page before and after the event firing can be made at any time after the firing of the event.

Reviewer: I think the assumptions:
1. should be made more proeminent, e.g. with an "Assumptions" section in the def.
2. should be copied in any rule using that definition, given that if the definition gives bad result, then the rule will also give bad result (probably together with a "conclusion" like "if this does not hold, then the def will produce nonsense and the rule will fail because of that without the SC failing" as we usually have.

(and I think that doing 1. also helps remembering to do 2.)

We don't really have assumptions in definitions so far, so there is no real agreement on how to handle them and no guidelines, so this is mostly personal opinion on how to handle them…

Author: I like the suggestion. Going to implement it :-)

---
title: Changes in content
key: changes-in-content
unambiguous: true
objective: true
input_aspects:
  - Accessibility tree
  - Audio output
  - CSS styling
  - DOM tree
---

> A _change in the content_ of a [web page][] caused by an [event][] occurs when, by comparing the [web page][] before and 1 minute after the event [firing][], at least one of the following occurs:

Reviewer: I'm considering
```suggestion
A _change in the content of a [web page][] caused by an [event][]_ occurs when, by comparing the [web page][] before and 1 minute after the event [firing][], at least one of the following occurs:
```
to emphasise that the event causing it is an important part of the definition (vs, say, time-based media).
Not super happy about this because the "of a web page" in the middle is diluting the intention of emphasising "change in the content caused by an event"…

Author: I think I got a way to address that

- **visible changes:** the rendered pixels change in any part of the document that is currently within the [viewport][] or that can be brought into the [viewport][] via scrolling; or
- **accessibility tree changes:** any state, property or event of a node representing an [accessible object][] of the [accessibility tree][] changes, or any node is inserted in, or removed from the [accessibility tree][]; or
- **audible changes:** the audio rendered by the [web page][] changes.

- If the [web page][] is rendering time-based media, rendered pixels and audio will be changing as part of the playback. The comparison in this instance should compare the pixels and audio that are rendered if the event is not fired, with the ones that are rendered if the event is fired.

**Assumptions:**

> - This definition assumes that the event causing the changes is the only origin of changes to the [web page][] and therefore the page does not change if there is no such event. If this is not the case, changes may be attributed to the wrong event.

Reviewer: I am still uneasy about the combination of these two.
I totally get the idea (I think) and agree with it, but the formulation seems contradictory…

It looks like we say on one hand what to do if there are other changes happening and on the other hand that no other changes are happening…

Of course, this is because the assumption is about "changes caused by something (DOM event or else in that case)" while the note above is about "ongoing changes that are not caused by anything anymore".

Given that the assumption is copied in the rules, I'd say it is the part that need clarification.

Maybe something like
> This definition/rule assumes that there are no changes in the content caused by another event.

Author: I see your point and I agree that it can be better formulated. I'll update based on your suggestion.

unambiguous: true
objective: true
input_aspects:
  - Accessibility tree
  - Audio output
  - CSS styling
  - DOM tree
---

> A _change in the content_ of a [web page][] caused by an event occurs when, by comparing the [web page][] before and after the event firing, at least one of the following occurs:

Reviewer: This does assume that firing the event is indeed causing the change.

If there is a page with audio playing non-stop (so, permanent change in content), then any event will be consider as a cause for this (because comparing the page before and after firing the event creates an **audible change**)

So, if there is a non-stop audio and a device motion event that does **not** cause any change in the content (e.g. it update a database in the backend), the definition will nonetheless consider the **audible change** as a "change in the content caused by" the device motion and the rule will fail while 2.5.4 doesn't (non-stop audio is bad for 1.4.2, of course).

Author: I shouldn't have used "event" because it was obvious that was going to be confused with dom events. I was using "event" with the generic meaning of "origin of the change" whatever that might be (most probably a dom event!)

Nevertheless, in your example, according to this definition, there would be no "change in content". There is a paragraph after the list explaining that when time-based media is at play, the comparison is not between what was before the event and what is after the event, but instead, between what would be if there was no event and what is when there is an event.

Reviewer: * "event": the use both of the word "event" and "firing" does make the confusion with DOM events likely. Not really important due to the rest.
* I have indeed missed the rest of the def 😥

> A _change in the content_ of a [web page][] caused by an event occurs when, by comparing the [web page][] before and after the event firing, at least one of the following occurs:

Reviewer: And timing strikes here: "after the event firing" => how long after? What happens if there is a delay between firing the event and causing the change? (that was a bit present before and becomes somewhat more obvious with this formulation…)
This is not too bad as not detecting a "change caused by an event" result in passing the rule (due to **no changes**) while failing the SC, but that does not break the fail-on-fail relationship.

Author: I also already had an assumption for this situation: "For the purpose of this definition, it is assumed that the changes happen instantly after the event firing and therefore the comparison between the page before and after the event firing can be made at any time after the firing of the event."

Reviewer: Missed that assumption indeed 😥

- **visible changes:** the rendered pixels change in any part of the document that is currently within the [viewport][] or that can be brought into the [viewport][] via scrolling; or
- **accessibility tree changes:** any state, property or event of a node representing an [accessible object][] of the [accessibility tree][] changes, or any node is inserted in, or removed from the [accessibility tree][]; or

> - **audible changes:** the audio rendered by the [web page][] changes.

Reviewer: I'm actually not sure whether "rendered" is the correct word to use for audio (or whether it is purely visual).

https://html.spec.whatwg.org/multipage/rendering.html
https://html.spec.whatwg.org/multipage/rendering.html#being-rendered
https://html.spec.whatwg.org/multipage/rendering.html#embedded-content-rendering-rules

This part is non normative, but the second link strongly suggests that we should only use "rendered" for visual stuff, not for audio.

Author: The W3C Web Audio API uses "audio rendering" has does almost everyone that works in the field :-)
But I agree that the HTML spec does not, so I'll change to "reproduced"

- **visible changes:** the rendered pixels change in any part of the document that is currently within the [viewport][] or that can be brought into the [viewport][] via scrolling; or
- **accessibility tree changes:** any state, property or event of a node representing an [accessible object][] of the [accessibility tree][] changes, or any node is inserted in, or removed from the [accessibility tree][]; or
- **audible changes:** the audio rendered by the [web page][] changes.

**Assumptions:**

> - If the [web page][] is rendering time-based media, rendered pixels and audio will be changing as part of the playback. The comparison in this instance should compare the pixels and audio that are rendered if the event is not fired, with the ones that are rendered if the event is fired.

Reviewer: I don't think this is an assumption.
As in, it's not something that we suppose and that make things go wrong if it doesn't hold.
It's a precision on how to handle a specific situation.

> - For the purpose of this definition, it is assumed that the event causing the changes is the only origin of changes to the [web page][] and therefore the page does not change if there is no such event.

Reviewer: ```suggestion
- This definition assumes that the event causing the changes is the only origin of changes to the [web page][] and therefore the page does not change if there is no such event. If this is not the case, changes may be attributed to the wrong event.
```

> - For the purpose of this definition, it is assumed that the changes happen instantly after the event firing and therefore the comparison between the page before and after the event firing can be made at any time after the firing of the event.

Reviewer: ```suggestion
- This definition assumes that the changes happen instantly after the event firing and therefore the comparison between the page before and after the event firing can be made at any time after the firing of the event. If there is a significant delay between firing the event and the changes happening, this definition may not detect them.
```
